Add lazy tensor python bindings

wconstab opened the pull request:
Differential Revision: D35032152

This adds a minimal set of python bindings for lazy tensor and the torchscript backend.  

It targets the APIs that are used by the `test_ts_opinfo.py` test (which it also lands, from lazy_tensor_staging, where it is [lazy_tensor_core/test/test_lazy.py](https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/test/test_lazy.py)).

We should land more python bindings obviously.  I just wanted to focus on a minimal set that can also be tested, and use it to agree on how we organize the bindings, then others could easily contribute bindings on top of this infrastructure.

cc @JackCaoG

alanwaketan commented on torch/csrc/lazy/python/init.cpp:
```diff
@@ -1,12 +1,607 @@
 #include <torch/csrc/lazy/python/python_util.h>
 
+
+// #include <c10/core/Device.h>
```
Any reasons why we commit so many lines of commented out code?

wconstab replied:
this isn't going to be landed this way.  I just wanted to bikeshed the layout of the files and the module conventions, then start to complete the bindings

also, sorry that this PR shows all the changes from the previous 2 diffs in the stack.  Hopefully you only looked at the changes from the top commit.

wconstab commented on test/lazy/test_ts_opinfo.py:
```diff
@@ -0,0 +1,161 @@
+
+            assert_allclose_rec((r_actual, r_exp))
+
+# TODO: after we move to master, add Lazy as a new Device here:
```
@Krovatkin what does this do? is this going to run a whole bunch of tests on the lazy device if we just wrote 'only_for="lazy"'?

Krovatkin replied:
yes

wconstab commented on test/lazy/test_ts_opinfo.py:
```diff
@@ -0,0 +1,161 @@
+            [conv_copy_out], [inp_copy, weight_copy, bias_copy], [grad_copy])
+
+        # TODO(whc) do we want to change how this API is exposed?
+        # jit_graph = torch._lazy._get_ltc_tensors_backend([bias_copy_grad])
```
@Krovatkin I assume this line is not important.  (i also assume you wrote it, maybe @desertfire did?)

Krovatkin replied:
yes you can remove it. When core folks fixed convolution to flow through one op, @desertfire removed the graph checks since convolutions become just another op

wconstab commented on torch/csrc/lazy/python/init.cpp:
```diff
@@ -1,12 +1,86 @@
+  torch::lazy::LazyGraphExecutor::Get()->SyncLiveTensorsGraph(&device, devices, wait);
+  torch::lazy::LazyGraphExecutor::Get()->MarkStep(device);
+
+  // TODO(whc) enable or delete this? (have we used it?)
```
As we land these bindings, i'm wondering if we should try to simplify by not having helper functions like 'StepMarker'.

Taking the full view of the mark step API flow, it looks like

1 torch._lazy.mark_step (python code in torch/_lazy/__init__.py)
2   -> torch._C._lazy._mark_step (binding code in lazy/python/init.cpp)
3      -> StepMarker (local helper function in lazy/python/init.cpp)
4        -> finally calls torch::lazy::blah APIs

(2) can always be squished into (3) but that can make gdb debugging more annoying.  But I also wonder if we can simplify to only have 3 layers, and generally flush stuff that would be in (2) either up into python in (1) or down into the core APIs (4).

Thoughts? @Krovatkin @alanwaketan

alanwaketan replied:
I think this is just Google's coding style to prefer wrapping a pile of code into a function just for readability not reusability.

Krovatkin replied:
I'm fine either way. 

> (2) can always be squished into (3) but that can make gdb debugging more annoying.

If users don't really need a python binding for 2), I'd suggest we go for it.

wconstab commented on torch/_lazy/__init__.py:
```diff
@@ -0,0 +1,25 @@
+
+
+def mark_step(device: torch.device = "lazy:0"):
+    # if xu.getenv_as('LTC_EMIT_STEPLOG', bool, False):
```
@JackCaoG here's a great example to discuss.  How do you envision the mark_step API working when torch-xla is part of torch::lazy?  Do you envision your users would ultimately `import torch._lazy.mark_step`, once you get past the stages of migration?

note: I did not intend to stop supporting `_run_step_closures()` but I might consider landing without it until we actually add some test coverage / usage scenarios of it.

Generally, my approach would be to only land the minimum bindings/apis we know we need and then keep adding them, rather than just land a bunch up front..

alanwaketan replied:
I remember the decision is to add the `GetDefaultDevice` API? It's just that we never find time to do that. Maybe we should add a comment above.

wconstab replied:
I think we actually do have a 'GetDefaultDevice' API for the backend, but we assume it is static after initialization time.  Maybe what I didn't want to do was have 'set/get current device' and have some state changing.

I was also thinking about how to improve our configuration flags and backend init APIs.  If we provide a good way to configure the backend configuration (python binding and then ENV override for each flag) could you use one of those flags to set default device type?  Or do you need something more than that?

alanwaketan replied:
This is the dummy stub we have in the staging branch: https://github.com/pytorch/pytorch/blob/8b8c69101c4da12df6dd6782906100a4f0e46779/lazy_tensor_core/lazy_tensor_core/csrc/init_python_bindings.cpp#L587.

JackCaoG commented:
I will take a look today

wconstab commented on torch/_lazy/metrics.py:
```diff
@@ -0,0 +1,8 @@
+import torch._C._lazy
+
+def reset():
```
Hadn't thought about this.

Right now we treat the counters' pybindings as just 'getters' (with the exception of reset).  We don't actually have any API for 'configuring' the counters, bc I believe they are just configured by static macros.

I am not sure if this system will work once we combine some in-tree and external-backend code, or we will have to adapt it to 'merge' the counters from external backend together with the ones from torch::lazy:: code.

Krovatkin commented on torch/_C/__init__.pyi.in:
```diff
@@ -22,6 +22,8 @@ import builtins
 from . import _nn as _nn
 from . import _onnx as _onnx
 from . import _VariableFunctions as _VariableFunctions
+from . import _lazy as _lazy
```
is this some weird python thing?
why can't we just omit `as _lazy`?

wconstab replied:
hmm, I have no idea, i just copied what others did.

Krovatkin commented on torch/_C/_lazy.pyi:
```diff
@@ -0,0 +1,7 @@
+from typing import List
+
+#defined in torch/csrc/lazy/python/init.cpp
```
nizzzeeeeeeee!

Krovatkin commented on torch/csrc/lazy/python/init.cpp:
```diff
@@ -1,12 +1,86 @@
-void initLazyBindings(PyObject* /* module */){
+
+struct NoGilSection {
+  NoGilSection() : state(PyEval_SaveThread()) {}
```
can we use `pybind11::gil_scoped_release` ? this is what pytorch seems to recommend

Krovatkin commented on torch/csrc/lazy/python/init.cpp:
```diff
@@ -1,12 +1,86 @@
+  // TODO(whc) enable or delete this? (have we used it?)
+  // bool debug_mode = lazy_tensors::sys_util::GetEnvBool("PT_LTC_DEBUG", false);
+  // if (C10_UNLIKELY(debug_mode)) {
+  //   std::string report = lazy_tensors::metrics::CreatePerformanceReport();
```
I'd say let's nuke it.
We have pretty decent python API for querying counters.

Krovatkin reviewed (approved):
:shipit:

antoniojkim commented on torch/_lazy/__init__.py:
```diff
@@ -0,0 +1,22 @@
+import torch._C._lazy
+
+
+def mark_step(device: str = "lazy:0", wait=False):
```
is the plan to expose all LTC python functions via `torch._lazy`? doesn't the underscore usually imply "private" and user shouldn't call? or is it eventually going to something like `torch.lazy` and the API exposed more similarly to how its currently done on the staging branch?

wconstab replied:
yea, torch._lazy signifies it is not a stable API, that's our convention.  It can't be torch.lazy until we (as a team) make a stronger commitment to long term support, etc.

antoniojkim commented on torch/csrc/lazy/python/init.cpp:
```diff
@@ -1,12 +1,65 @@
+void initLazyBindings(PyObject* module){
+  auto m = py::handle(module).cast<py::module>();
+  auto lazy = m.def_submodule("_lazy");
+  auto lazy_ts_backend = m.def_submodule("_lazy_ts_backend");
```
In fact, this might be what you intended all along based on its usage in `test_op_info.py`:
https://github.com/pytorch/pytorch/pull/74508/files#diff-222fb8e63ff517f76308939550ca744de90dd6ea5706e2afd3f72db019df9b90R24

wconstab replied:
yea its a good question;
I tried to make it clean from the actual user API side, which is how it looks when called from test_ts_opinfo.py.

generally the c++ bindings are flatter all over pytorch and then structurized at the glue-layer where .py files re-export them in a proper namespace (the stuff happening in torch/_lazy/*.py), so I didn't think it mattered too much.  

I might be convinced to change it though.  Wdyt @Krovatkin ?  One annoying thing is I would have to then define a 'submodule' correctly in the .pyi file that provides the mypy formatting, but i'm sure that's  not too hard

wconstab committed:
Add lazy tensor python bindings (#74508)

Summary:
This adds a minimal set of python bindings for lazy tensor and the torchscript backend.

It targets the APIs that are used by the `test_ts_opinfo.py` test (which it also lands, from lazy_tensor_staging, where it is [lazy_tensor_core/test/test_lazy.py](https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/test/test_lazy.py)).

We should land more python bindings obviously.  I just wanted to focus on a minimal set that can also be tested, and use it to agree on how we organize the bindings, then others could easily contribute bindings on top of this infrastructure.

cc JackCaoG

Pull Request resolved: https://github.com/pytorch/pytorch/pull/74508

Differential Revision: D35032152

Pulled By: wconstab

fbshipit-source-id: 0b6ff9b3a82cd63801dac3d267c1909bbf1f4de4
